### PR TITLE
Improve cluster failure testing

### PR DESCRIPTION
**Pull request by rickardoberg**
NetworkSender/NetworkReceiver now uses pausing instead of start/stop for fail testing.
Refactored ClusterClient into factory+facade
Always use "127.0.0.1" for localhost in cluster testing, to avoid the 127.0.1.1 issue.
Plus other minor code cleanups.

### CODE REVIEW

File: enterprise/ha/src/test/java/org/neo4j/ha/TransactionConstraintsIT.java [hunk 1]
```diff
@@ -312,9 +318,12 @@ public Node call() throws Exception
         thread2.close();
     }
 
+    @RepeatRule.Repeat(times=1000)
     @Test
     public void createdSchemaConstraintsMustBeRetainedAcrossModeSwitches() throws Throwable
     {
+        System.out.println(repeatRule.getCount());
+
```
Comment (chrisvest) [inline enterprise/ha/src/test/java/org/neo4j/ha/TransactionConstraintsIT.java]: Nope.

File: community/neo4j-slf4j/src/main/java/org/neo4j/logging/slf4j/Slf4jLog.java [hunk 2]
```diff
@@ -239,4 +239,9 @@ public void bulk( Consumer<Log> consumer )
             consumer.accept( this );
         }
     }
+
+    private String convertFormat(String format)
```
Comment (digitalstain) [inline community/neo4j-slf4j/src/main/java/org/neo4j/logging/slf4j/Slf4jLog.java:243]: Why is this necessary?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/AcceptorState.java [hunk 3]
```diff
@@ -66,6 +66,16 @@ public AcceptorState handle( AcceptorContext context,
                             AcceptorMessage.PrepareState incomingState = message.getPayload();
                             InstanceId instanceId = new InstanceId( message );
 
+/*
```
Comment (digitalstain) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/AcceptorState.java]: Is this necessary to be included?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/InMemoryAcceptorInstanceStore.java [hunk 4]
```diff
@@ -73,7 +73,7 @@ public void lastDelivered( InstanceId instanceId )
     {
         for ( long i = lastDeliveredInstanceId; i <= instanceId.getId(); i++ )
```
Comment (digitalstain) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/InMemoryAcceptorInstanceStore.java]: Better remove this all together or perhaps replace it with a comment explaining what used to happen here to maintain a more accessible history?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/LearnerState.java [hunk 5]
```diff
@@ -252,7 +254,7 @@ public LearnerState handle( LearnerContext context,
 
                                 context.setLastKnownLearnedInstanceInCluster(
                                         catchUpTo,
-                                        context.getIdForUri( new URI( message.getHeader( Message.FROM ) ) ) );
+                                        new org.neo4j.cluster.InstanceId( Integer.parseInt( message.getHeader( Message.INSTANCE_ID ) )));
```
Comment (digitalstain) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/LearnerState.java:257]: This will cause issues if we run mixed version clusters - the old version instances will not contain the INSTANCE_ID header and will cause failures. We should probably have this as an if check that falls back to URI based lookups as before.

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/heartbeat/HeartbeatState.java [hunk 6]
```diff
@@ -229,8 +229,8 @@ public HeartbeatState handle( HeartbeatContext context,
                             context.getInternalLog( HeartbeatState.class )
                                     .debug( "Received suspicions as " + suspicions );
 
-                            URI from = new URI( message.getHeader( Message.FROM ) );
-                            InstanceId fromId = context.getIdForUri( from );
+                            InstanceId fromId = new InstanceId(Integer.parseInt(message.getHeader( Message.INSTANCE_ID )));
```
Comment (digitalstain) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/heartbeat/HeartbeatState.java:232]: Same comment here as before about backwards compatibility. This should also check against URIs

File: enterprise/cluster/src/main/java/org/neo4j/cluster/statemachine/StateTransitionLogger.java [hunk 7]
```diff
@@ -79,6 +83,19 @@ public void stateTransition( StateTransition transition )
...
+                if (payload instanceof Payload )
+                {
+                    try
+                    {
+                        payload = atomicBroadcastSerializer.receive( (Payload) payload);
+                    }
+                    catch ( Throwable e )
+                    {
+                        // Ignore
+                        e.printStackTrace();
```
Comment (digitalstain) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/statemachine/StateTransitionLogger.java]: Are we sure printing the stack trace is the proper action here? If this is ignoreable, then nothing should be printed - if we are interested in this, then we should print int out in a more structured way?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/client/Clusters.java [hunk 8]
```diff
@@ -203,14 +201,7 @@ public boolean isFullHaMember()
...
         {
-            try
-            {
-                return InetAddress.getLocalHost().getHostAddress();
-            }
-            catch ( UnknownHostException e )
-            {
-                throw new RuntimeException( e );
-            }
+            return "127.0.0.1";
```
Comment (digitalstain) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/client/Clusters.java:204]: Why do we need this particular address? Shouldn't it be sufficient to have a consistent string returned? If that's the case, then why do we need to change this code, since it seems to be returning a consistent string anyway? I fear this may be masking a problem where somewhere else in the network stack we are making a different assumption about our available interfaces. That doesn't necessarily mean that this change is wrong, but we may be missing a deeper bug.

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerState.java [hunk 9]
```diff
@@ -83,7 +83,8 @@ public ProposerState handle( ProposerContext context,
                                     + " rejected from " + message.getHeader( Message.FROM ) + " with ballot "
                                     + rejectPropose.getBallot() );
 
-                            if ( instance.isState( PaxosInstance.State.p1_pending ) )
+                            if ( instance.isState( PaxosInstance.State.p1_pending ) &&
```
Comment (digitalstain) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerState.java:86]: This if statement is quite opaque - some documentation would be nice

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/context/AcceptorContextImpl.java [hunk 10]
```diff
@@ -64,10 +67,16 @@ public void leave()
         instanceStore.clear();
     }
 
+    @Override
+    public boolean isClosed( InstanceId instanceId )
+    {
+        return false; // return paxosInstances.getPaxosInstance( instanceId );
```
Comment (digitalstain) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/context/AcceptorContextImpl.java]: I think that comment could be replaced with something more explanatory

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityModeSwitcher.java [hunk 11]
```diff
@@ -362,7 +363,6 @@ public void run()
                     }
 
                     modeSwitcherExecutor.schedule( this, 5, TimeUnit.SECONDS );
```
Comment (digitalstain) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityModeSwitcher.java:364]: Not directly related to this PR, but isn't this rescheduling wrong?

File: enterprise/server-enterprise/src/test/java/org/neo4j/server/enterprise/StandaloneClusterClientIT.java [hunk 12]
```diff
@@ -303,6 +306,7 @@ private Process startStandaloneClusterClientProcess( File configFile, Map<String
         {
             args.add( "-" + entry.getKey() + "=" + entry.getValue() );
         }
+        args.add("-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=5005");
```
Comment (digitalstain) [inline enterprise/server-enterprise/src/test/java/org/neo4j/server/enterprise/StandaloneClusterClientIT.java]: Should this be left here?

**Comment by digitalstain** [comment]
Should there perhaps be some additional test cases for the Acceptor bug fixed? I don't think it's covered by the set of tests we currently have in place.

**Comment by digitalstain** [comment]
Shouldn't this patch be issued against 2.2?

File: community/neo4j-slf4j/src/main/java/org/neo4j/logging/slf4j/Slf4jLog.java (hunk 2 above)
Comment (rickardoberg) [inline community/neo4j-slf4j/src/main/java/org/neo4j/logging/slf4j/Slf4jLog.java:243]: The conversion? Because SLF4J and String.format uses different placeholder markers (%s vs {})

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerMessage.java [hunk 13]
```diff
@@ -61,9 +64,22 @@ public Object getValue()
...
+            Object toStringValue = value;
+            if (toStringValue instanceof Payload )
+            {
+                try
+                {
+                    toStringValue = new AtomicBroadcastSerializer( new ObjectStreamFactory(), new ObjectStreamFactory() ).receive( (Payload) toStringValue);
+                }
+                catch ( Throwable e )
+                {
+                    e.printStackTrace();
```
Comment (digitalstain) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerMessage.java:76]: Perhaps log it instead?

**Comment by davidegrohmann** [comment]
@rickardoberg this needs a rebase

**Comment by benbc** [comment]
Please put this fix into 2.2

**Comment by tinwelint** [comment]
Can this PR be closed now when there's #4804 for 2.2?

**Comment by rickardoberg** [comment]
No, the code is very different, and the 2.2 one cannot be forward merged. They need to be merged separately.

**Comment by davidegrohmann** [comment]
@rickardoberg can we apply this changes when forward merging #4804? Moreover you have an extra commit "Make HAGDB GCable" in that PR which I guess it should be forward merged.

Btw this needs a rebase and there are still quite a few comments not addressed yet.

**Comment by rickardoberg** [comment]
@davidegrohmann #4804 should not really be forward merged. The two PRs are different, because of differences between 2.2 and 2.3 when it comes to logging and the module system (in particular). I've added the GC stuff to this PR as well.

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/LearnerState.java (hunk 5 above)
Comment (rickardoberg) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/LearnerState.java:257]: We have had INSTANCE_ID since 1.9, so this is not relevant IMHO.

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/heartbeat/HeartbeatState.java (hunk 6 above)
Comment (rickardoberg) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/heartbeat/HeartbeatState.java:232]: See above

File: enterprise/cluster/src/main/java/org/neo4j/cluster/client/Clusters.java (hunk 8 above)
Comment (rickardoberg) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/client/Clusters.java:204]: getHostAddress can return 127.0.0.1, 127.0.1.1 or the host IP. Setting it to a fixed 127.0.0.1 makes sure we avoid issues where different instances get different results. And this is only for testing anyway, so does not impact customers.

File: enterprise/cluster/src/main/java/org/neo4j/cluster/StateMachines.java [hunk 14]
```diff
@@ -190,6 +190,8 @@ public void run()
                                 }
                                 else
                                 {
+//                                    message.copyHeadersTo( outgoingMessage, Message.INSTANCE_ID );
+
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/StateMachines.java:194]: to be removed?

File: community/neo4j-slf4j/src/main/java/org/neo4j/logging/slf4j/Slf4jLogProvider.java [hunk 15]
```diff
@@ -19,24 +19,38 @@
...
 /**
  * A {@link LogProvider} that forwards log events to SLF4J
  */
 public class Slf4jLogProvider implements LogProvider
 {
+    private ILoggerFactory loggerFactory;
+
+    public Slf4jLogProvider()
+    {
+        this( LoggerFactory.getILoggerFactory());
```
Comment (davidegrohmann) [inline community/neo4j-slf4j/src/main/java/org/neo4j/logging/slf4j/Slf4jLogProvider.java:37]: code style

File: enterprise/cluster/src/main/java/org/neo4j/cluster/NetworkedServerFactory.java [hunk 16]
```diff
@@ -140,7 +141,8 @@ public void listeningAt( URI me )
                 protocolServer.listeningAt( me );
                 if ( logger == null )
                 {
-                    logger = new StateTransitionLogger( logProvider );
+                    logger = new StateTransitionLogger(  logProvider,
+                            new AtomicBroadcastSerializer(objectInputStreamFactory, objectOutputStreamFactory));
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/NetworkedServerFactory.java:145]: code style

File: enterprise/cluster/src/main/java/org/neo4j/cluster/com/NetworkReceiver.java [hunk 17]
```diff
@@ -212,22 +219,25 @@ public void addMessageProcessor( MessageProcessor processor )
 
     public void receive( Message message )
     {
-        for ( MessageProcessor processor : processors )
+        if (!paused)
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/com/NetworkReceiver.java:222]: invert if for readability?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/com/NetworkSender.java [hunk 18]
```diff
@@ -215,18 +217,26 @@ public void process( final List<Message<? extends MessageType>> messages )
     @Override
     public boolean process( Message<? extends MessageType> message )
     {
-        if ( message.hasHeader( Message.TO ) )
-        {
-            send( message );
-        }
-        else
+        if (!paused)
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/com/NetworkSender.java:220]: invert if?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/AtomicBroadcastSerializer.java [hunk 19]
```diff
@@ -33,6 +33,11 @@
     private ObjectInputStreamFactory objectInputStreamFactory;
     private ObjectOutputStreamFactory objectOutputStreamFactory;
 
+    public AtomicBroadcastSerializer()
+    {
+        this(new ObjectStreamFactory(), new ObjectStreamFactory());
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/AtomicBroadcastSerializer.java:38]: code style

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/AcceptorMessage.java [hunk 20]
```diff
@@ -145,7 +148,20 @@ public int hashCode()
...
+            if (toStringValue instanceof Payload )
+            {
+                try
+                {
+                    toStringValue = new AtomicBroadcastSerializer( new ObjectStreamFactory(), new ObjectStreamFactory() ).receive( (Payload) toStringValue);
+                }
+                catch ( Throwable e )
+                {
+                    e.printStackTrace();
+                }
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/AcceptorMessage.java:161]: move this to `Paylod.toString`?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/PaxosInstance.java [hunk 21]
```diff
@@ -191,10 +199,43 @@ public PaxosInstance snapshot(PaxosInstanceStore store)
...
+
+            return "[id:" + id +
+                   " state:" + state.name() +
+                   " b:" + ballot +
+                   " v1:" + toStringValue1 +
+                   " v2:" + toStringValue2 + "]";
+        }
+        catch ( Throwable e )
+        {
+            return "";
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/PaxosInstance.java:238]: perhaps just skip the non-serializable values only?

File: enterprise/cluster/src/test/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerStateTest.java [hunk 22]
```diff
@@ -90,6 +92,36 @@ public void ifProposingWithClosedInstanceThenRetryWithNextInstance() throws Thro
     }
 
     @Test
+    public void something() throws Throwable
```
Comment (davidegrohmann) [inline enterprise/cluster/src/test/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerStateTest.java]: something?

File: enterprise/cluster/src/test/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerStateTest.java [hunk 23]
```diff
@@ -90,6 +92,36 @@ public void ifProposingWithClosedInstanceThenRetryWithNextInstance() throws Thro
...
+                        create( "http://something3" ) ) ) );
+
+        Message message = Message.to( ProposerMessage.promise, create( "http://something1" ), new ProposerMessage.PromiseState( 2001, acceptorValue ) );
+        message.setHeader( org.neo4j.cluster.protocol.atomicbroadcast.multipaxos.InstanceId.INSTANCE, instanceId.toString() );
+
+        MessageHolder mockHolder = mock( MessageHolder.class );
+        ProposerState.proposer.handle(context, message, mockHolder);
+
+
+
```
Comment (davidegrohmann) [inline enterprise/cluster/src/test/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/ProposerStateTest.java:121]: no assertions?

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityModeSwitcher.java [hunk 24]
```diff
@@ -351,6 +351,7 @@ public void run()
                     ClusterClient clusterClient = HighAvailabilityModeSwitcher.this.clusterClient;
                     try
                     {
+                        // TODO I doubt this actually works
```
Comment (davidegrohmann) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityModeSwitcher.java:354]: improve comment with explanation why this doesn't work?

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterTopologyChangesIT.java [hunk 25]
```diff
@@ -78,6 +79,9 @@
     @Rule
     public final CleanupRule cleanup = new CleanupRule();
 
+    @Rule
+    public final RepeatRule repeat = new RepeatRule();
+
```
Comment (davidegrohmann) [inline enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterTopologyChangesIT.java:84]: to be removed?

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterTopologyChangesIT.java [hunk 26]
```diff
@@ -90,20 +94,34 @@ public void setup() throws Exception
                 .startCluster();
     }
 
+    @After
+    public void cleanup()
+    {
+        cluster = null;
+        System.gc();
```
Comment (davidegrohmann) [inline enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterTopologyChangesIT.java:101]: really needed?

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterTopologyChangesIT.java [hunk 27]
```diff
@@ -90,20 +94,34 @@ public void setup() throws Exception
...
+    }
+    
     @Test
     public void masterRejoinsAfterFailureAndReelection() throws Throwable
     {
+        System.out.println(repeat.getCount());
+
```
Comment (davidegrohmann) [inline enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterTopologyChangesIT.java:108]: to be removed?

File: enterprise/cluster/src/main/java/org/neo4j/cluster/client/ClusterClientModule.java [hunk 28]
```diff
@@ -0,0 +1,292 @@
...
+    public final ProtocolServer server;
+    public final InMemoryAcceptorInstanceStore acceptorInstanceStore;
+    public final MessageTimeoutStrategy timeoutStrategy;
+    public final NetworkReceiver receiver;
+
+    public ClusterClientModule( LifeSupport parentLife, Dependencies dependencies, final Monitors monitors, final
+    Config config, LogService logService, ElectionCredentialsProvider electionCredentialsProvider )
+    {
+        this.life = new LifeSupport();
+        if (parentLife != null)
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/client/ClusterClientModule.java:87]: code style

File: enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/InMemoryAcceptorInstanceStore.java [hunk 29]
```diff
@@ -51,6 +56,11 @@ public AcceptorInstance getAcceptorInstance( InstanceId instanceId )
         {
             instance = new AcceptorInstance();
             instances.put( instanceId, instance );
+            if (!currentInstances.offer( instanceId ))
```
Comment (davidegrohmann) [inline enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/InMemoryAcceptorInstanceStore.java:59]: code style
